op-node/l1/receipts.go

```diff
@@ -78,9 +83,45 @@ func fetchReceipts(ctx context.Context, block eth.BlockID, receiptHash common.Ha
 	// Sanity-check: external L1-RPC sources are notorious for not returning all receipts,
 	// or returning them out-of-order. Verify the receipts against the expected receipt-hash.
 	hasher := trie.NewStackTrie(nil)
-	computed := types.DeriveSha(types.Receipts(receipts), hasher)
-	if receiptHash != computed {
-		return nil, fmt.Errorf("failed to fetch list of receipts: expected receipt root %s but computed %s from retrieved receipts", receiptHash, computed)
+	computed := types.DeriveSha(types.Receipts(rf.receipts), hasher)
+	if rf.receiptHash != computed {
+		return nil, fmt.Errorf("failed to fetch list of receipts: expected receipt root %s but computed %s from retrieved receipts", rf.receiptHash, computed)
 	}
-	return receipts, nil
+	return rf.receipts, nil
+}
+
+// NewReceiptsFetcher creates a receipt fetcher that can iteratively fetch the receipts matching the given t
+func NewReceiptsFetcher(block eth.BlockID, receiptHash common.Hash, txs types.Transactions, getBatch batchCallContextFn, batchSize int) (eth.ReceiptsFetcher, error) {
+	if len(txs) == 0 {
+		if receiptHash != types.EmptyRootHash {
+			return nil, fmt.Errorf("no transactions, but got non-empty receipt trie root: %s", receiptHash)
+		}
+		return eth.FetchedReceipts(nil), nil
+	}
+	if len(txs) < batchSize {
+		batchSize = len(txs)
+	}
+	if batchSize < 1 {
+		batchSize = 1
+	}
+
+	receipts := make([]*types.Receipt, len(txs))
+	receiptRequests := make([]rpc.BatchElem, len(txs))
+	for i := 0; i < len(txs); i++ {
+		receipts[i] = new(types.Receipt)
+		receiptRequests[i] = rpc.BatchElem{
```

Comment:
I think it could be nice to have a helper function here that creates the `rpc.BatchElem`, something like `NewBatchGetTransactionReceipt(common.Hash)`. We could then have a library of these constructor functions and very cleanly build whatever sort of batch calls that we want

Reply:
Agreed, the RPC iterative batch fetching is a nice util, also for L2 RPC. Would like to minimize changes in this PR though, we can move it to a common package + polish for more general usage in a later PR.